lib/pulse_sensor.c

```diff
@@ -0,0 +1,332 @@
+			return NO_NEW_VALUE;
+		}
+	}
+
+	if (pulseSensor[sensorNumber].DMALastReadValue == valueInQuestion) {	// Check if the last read value is the same as the current
+		if (currentTick - pulseSensor[sensorNumber].lastDMAReadValueTimeMs >= pulseSensor[sensorNumber].dmaStoppedTimeoutMS){	// Check if we haven't changed values in a while which might mean we're stopped
+			pulseSensor[sensorNumber].stopped = true;
+
+			// Clear buffer so any non-zero values will be quickly identified that the car is moving again. This will also reset the DMA position.
+			clear_buffer_and_reset_dma(sensorNumber);
+
+			pulseSensor[sensorNumber].DMALastReadValue = 0;
+			pulseSensor[sensorNumber].lastDMAReadValueTimeMs = currentTick;
+			*(pulseSensor[sensorNumber].resultStoreLocation) = 0;
+
+#ifdef DEBUG_PS
+			printf("=== STOPPED ===\n");
+#endif
+
+			return STOPPED;
+		}
+		// Value is the same as the last but it hasn't been enough time to declare stale, hop out of here for now because there's nothing new to process
+		return NO_NEW_VALUE;
+	}
+
+
+	// ===== Passed All Breakpoint Checks =====
+	pulseSensor[sensorNumber].DMALastReadValue = valueInQuestion;
+	pulseSensor[sensorNumber].lastDMAReadValueTimeMs = currentTick;
+
+	// Declare amount of samples and evaluate initial size based on if we're using variable sampling. If we are we'll determine it again after some deltas are evaluated
+	U16 numSamples = pulseSensor[sensorNumber].useVariableSpeedSampling ? pulseSensor[sensorNumber].minSamples + 1 : MAX_DELTAS;
+
+	// Calculate the deltas between each of the time values and store them in deltaTotal
+	U32 deltaTotal = 0; // Gets big adding up all U32 deltas but not big enough to need U64 before stopped if like <1 full second
+	U32 lastDelta = 0;
+	U32 last2ndDelta = 0;
+	U16 numDeltas = 0;
+	U32 delta = 0;
+
+	// Begin by going through the first minimum amount samples to determine how many we should read, and then find the average delta while rejecting any dropped DMA values
+	for (U16 i = 0; i < numSamples; i++)
+	{
+		U32 value1 = pulseSensor[sensorNumber].buffer[(DMACurrentPosition - 1 - i + IC_BUF_SIZE) % IC_BUF_SIZE];
+		U32 value2 = pulseSensor[sensorNumber].buffer[(DMACurrentPosition - 2 - i + IC_BUF_SIZE) % IC_BUF_SIZE];
+		if (value1 != 0 && value2 != 0) { // Check if 0s because of previously empty buffer, because otherwise a value would simply be a time amount
+			// Check if the difference is so small it can be considered a duplicate of the last
+			if (abs(value2 - value1) > DUPLICATE_VALUE_TICK_DIFFERENCE) {
+				if (value1 < value2) {
+					delta = ((1 << pulseSensor[sensorNumber].timerSize) | value1) - value2; // Buffer roll-over (timer resets to 0) protection
+				} else {
+					delta = value1 - value2; // Buffer roll-over (timer resets to 0) protection
+				}
+#ifdef DEBUG_PS
+				// Store deltas in an array for  if we're debugging
+				deltaList[numDeltas] = delta;
```

Comment:
Why put this in the last value of the deltaList?

Reply:
numDeltas is a counter that counts up when its checking deltas and the 2 numbers its looking at are not 0 or duplicate values, so we know how many valid deltas there are (to divide the total by) and in the array it will just make there be no gaps in the array.